controllers/awsmachine_controller_unit_test.go

```diff
@@ -17,25 +17,1507 @@ limitations under the License.
+				expectConditions(g, ms.AWSMachine, []conditionAssertion{{conditionType: infrav1.InstanceReadyCondition, status: corev1.ConditionUnknown}})
+			})
+			t.Run("security Groups succeed", func(t *testing.T) {
+				getCoreSecurityGroups := func(t *testing.T, g *WithT) {
+					t.Helper()
+
+					ec2Svc.EXPECT().GetInstanceSecurityGroups(gomock.Any()).
+						Return(map[string][]string{"eid": {}}, nil)
+					secretSvc.EXPECT().UserData(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).Times(1)
+					ec2Svc.EXPECT().GetCoreSecurityGroups(gomock.Any()).Return([]string{}, nil)
+					secretSvc.EXPECT().Create(gomock.Any(), gomock.Any()).Return("test", int32(1), nil).Times(1)
+				}
+				t.Run("should reconcile security groups", func(t *testing.T) {
+					g := NewWithT(t)
+					awsMachine := getAWSMachine()
+					setup(t, g, awsMachine)
+					defer teardown(t, g)
+					instanceCreate(t, g)
+					getCoreSecurityGroups(t, g)
+
+					ms.AWSMachine.Spec.AdditionalSecurityGroups = []infrav1.AWSResourceReference{
+						{
+							ID: pointer.StringPtr("sg-2345"),
+						},
+					}
+					ec2Svc.EXPECT().UpdateInstanceSecurityGroups(instance.ID, []string{"sg-2345"})
+
+					_, _ = reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+					expectConditions(g, ms.AWSMachine, []conditionAssertion{{conditionType: infrav1.SecurityGroupsReadyCondition, status: corev1.ConditionTrue}})
+				})
+
+				t.Run("should not tag anything if there's not tags", func(t *testing.T) {
+					g := NewWithT(t)
+					awsMachine := getAWSMachine()
+					setup(t, g, awsMachine)
+					defer teardown(t, g)
+					instanceCreate(t, g)
+					getCoreSecurityGroups(t, g)
+
+					ec2Svc.EXPECT().UpdateInstanceSecurityGroups(gomock.Any(), gomock.Any()).Times(0)
+					if _, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs); err != nil {
+						_ = fmt.Errorf("reconcileNormal reutrned an error during test")
+					}
+				})
+
+				t.Run("should tag instances from machine and cluster tags", func(t *testing.T) {
+					g := NewWithT(t)
+					awsMachine := getAWSMachine()
+					setup(t, g, awsMachine)
+					defer teardown(t, g)
+					instanceCreate(t, g)
+					getCoreSecurityGroups(t, g)
+
+					ms.AWSMachine.Spec.AdditionalTags = infrav1.Tags{"kind": "alicorn"}
+					cs.AWSCluster.Spec.AdditionalTags = infrav1.Tags{"colour": "lavender"}
+
+					ec2Svc.EXPECT().UpdateResourceTags(
+						gomock.Any(),
+						map[string]string{
+							"kind": "alicorn",
+						},
+						map[string]string{},
+					).Return(nil).Times(2)
+
+					ec2Svc.EXPECT().UpdateResourceTags(
+						PointsTo("myMachine"),
+						map[string]string{
+							"colour": "lavender",
+							"kind":   "alicorn",
+						},
+						map[string]string{},
+					).Return(nil)
+
+					_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+					g.Expect(err).To(BeNil())
+				})
+				t.Run("should tag instances volume tags", func(t *testing.T) {
```

Comment:
this test is specifically to check for volume tags, i remember since this was added by me in [this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2463) PR. other testcase covers general additionalTags.

Reply:
I don't see any difference other than tags are having different values.